gossip/gossip.go

```diff
@@ -118,10 +133,32 @@ func (g *Gossiper) Sync(subnet, hash id.Hash) {
 		Type:    wire.Pull,
 		Data:    marshaledPullV1,
 	}
-	g.sendToSubnet(subnet, msg)
+
+	// Before sending the message to the subnet, store a responder channel in
+	// the gossiper so we can receive the response.
+	g.syncRespondersMu.Lock()
+
+	responder := make(chan []byte, 1)
+	if len(g.syncResponders[hash]) == 0 {
+		// Only send the message to the subnet if the hash does not exist. This
+		// ensures we do not send it multiple times.
+		g.sendToSubnet(subnet, msg)
+	}
+	g.syncResponders[hash] = append(g.syncResponders[hash], responder)
+
+	// Do not defer the unlocking of the mutex, because the next statement could
+	// block for a substantial amount of time.
+	g.syncRespondersMu.Unlock()
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case response := <-responder:
+		return response, nil
+	}
```

Comment:
```suggestion
	defer func() {
		g.syncRespondersMu.Lock()
		defer g.syncRespondersMu.Unlock()
		delete(g.syncResponders, hash)
	}()

	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case response := <-responder:
		return response, nil
	}
```

Reply:
What if there are other workers still waiting on the response? e.g. if we call `Sync` twice concurrently, one with a 1 second context and the other with a 1 minute context. If the response takes longer than a second, when the first context expires, the list of responders will be removed and the gossiper will no longer have access to the other responder that may still be waiting.

Reply:
True, in which case instead of `delete(g.syncResponders, hash)` we should do
```suggestion
	defer func() {
		g.syncRespondersMu.Lock()
		defer g.syncRespondersMu.Unlock()
		for i, ch := range g.syncResponders[hash] {
			if ch == responder { // Equality of channels is ok, according to the Go specification.
				g.syncResponders[hash][i] = g.syncResponders[hash][len(g.syncResponders[hash])-1]
				g.syncResponders[hash] = g.syncResponders[hash][:len(g.syncResponders[hash])-1]
				break
			}
		}
		if len(g.syncResponders[hash]) == 0 {
			delete(g.syncResponders, hash)
		}
	}()
	
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case response := <-responder:
		return response, nil
	}
```